### PR TITLE
[`utils`] add merge_lora utility function

### PR DESCRIPTION
**Pull request by younesbelkada**
# What does this PR do?

This PR introduces a new utility function `merge_lora` to conveniently merge lora layers into the base model, so that users can use this model as a standalone `transformers` model

Added also some tests (that are currently failing)

cc @pacman100

### CODE REVIEW

**Comment by younesbelkada** [comment]
This PR is now ready for review! all tests are passing

File: tests/test_peft_model.py [hunk 1]
```diff
@@ -24,23 +24,27 @@
...
+    merge_lora,
     prepare_model_for_int8_training,
 )
 
 from .testing_common import PeftTestConfigManager
 
 
-# This has to be in the order: model_id, lora_kwargs, prefix_tuning_kwargs, prompt_encoder_kwargs, prompt_tuning_kwargs
+# This has to be in the order: model_id, lora_kwargs, prefix_tuning_kwargs, prompt_encoder_kwargs,
 PEFT_DECODER_MODELS_TO_TEST = [
```
Comment (younesbelkada) [inline tests/test_peft_model.py:35]:
```suggestion
PEFT_DECODER_MODELS_TO_TEST = [
```

File: src/peft/tuners/lora.py [hunk 2]
```diff
@@ -233,6 +238,34 @@ def enable_adapter_layers(self):
...
+        This method merges the LoRa layers into the base model. This is needed if someone wants to use the base model
+        as a standalone model.
+        """
+        if self.config.model_type == "gpt2":
+            raise ValueError("GPT2 models are not supported for merging LORA layers")
+
+        key_list = [key for key, _ in self.model.named_modules() if "lora" not in key]
+        for key in key_list:
+            parent, target, target_name = self._get_submodules(key)
+            if isinstance(target, (Linear, MergedLinear)):
```
Comment (pacman100) [inline src/peft/tuners/lora.py:252]:
```suggestion
            if isinstance(target, LoraLayer):
```

File: src/peft/tuners/lora.py [hunk 3]
```diff
@@ -233,6 +238,34 @@ def enable_adapter_layers(self):
     def disable_adapter_layers(self):
         self._set_adapter_layers(enabled=False)
 
+    def merge_and_unload(self):
```
Comment (pacman100) [inline src/peft/tuners/lora.py:241]: This would not work when the model is loaded in 8bit. An assertion error for the same would be helpful.

**Review by pacman100 — COMMENTED** [review]
Thank you @younesbelkada for adding this feature which resolves many issues related to merging the lora params back into the base model without any wrappers on it.✨

Left a couple of comments and suggestions

File: src/peft/tuners/lora.py (hunk 3 above)
Comment (younesbelkada) [reply src/peft/tuners/lora.py:241]: Thanks for the pointer! added a check for that

File: src/peft/tuners/lora.py [hunk 4]
```diff
@@ -233,6 +238,37 @@ def enable_adapter_layers(self):
...
+                new_module = torch.nn.Linear(target.in_features, target.out_features, bias=bias)
+
+                # manually merge if not merged
+                if not target.merged:
+                    # merge weights per: https://arxiv.org/pdf/2106.09685.pdf / page 4
+                    if target.r > 0:
+                        target.weight.data += (
+                            transpose(target.lora_B.weight @ target.lora_A.weight, target.fan_in_fan_out)
+                            * target.scaling
+                        )
```
Comment (pacman100) [inline src/peft/tuners/lora.py:266]:
```suggestion
                        ).to(target.weight.dtype)
```

Comment (pacman100) [reply src/peft/tuners/lora.py:266]: if the model is loaded in fp16/bf16, then it would be better to first convert the lora matrix into that dtype.

**Review by pacman100 — COMMENTED** [review]
Final comments/suggestions

File: src/peft/tuners/lora.py (hunk 4 above)
Comment (younesbelkada) [reply src/peft/tuners/lora.py]: Yes indeed!

**Review by pacman100 — APPROVED** [review]

**Pull request merged**